ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,412 @@
+	/// Original signed transaction hash (used as a source for private tx)
+	pub tx_hash: H256,
+	/// Current status of the private transaction
+	pub status: PrivateTxStatus,
+	/// Creation timestamp
+	pub creation_timestamp: u64,
+	/// List of validations
+	pub validators: Vec<ValidatorLog>,
+	/// Timestamp of the resulting public tx deployment
+	pub deployment_timestamp: Option<u64>,
+	/// Hash of the resulting public tx
+	pub public_tx_hash: Option<H256>,
+}
+
+/// Wrapper other JSON serializer
+pub trait LogsSerializer: Send + Sync + 'static {
+	/// Read logs from the source
+	fn read_logs(&self) -> Result<Vec<TransactionLog>, Error>;
+
+	/// Write all logs to the source
+	fn flush_logs(&self, logs: &HashMap<H256, TransactionLog>) -> Result<(), Error>;
+}
+
+pub struct FileLogsSerializer {
+	logs_dir: Option<PathBuf>,
+}
+
+impl FileLogsSerializer {
+	pub fn new(logs_dir: Option<String>) -> Self {
+		FileLogsSerializer {
+			logs_dir: logs_dir.map(|dir| PathBuf::from(dir)),
+		}
+	}
+
+	fn open_file(&self, to_create: bool) -> Result<Option<File>, Error> {
+		let log_file = match self.logs_dir {
+			Some(ref path) => {
+				let mut file_path = path.clone();
+				file_path.push("private_tx.log");
+				match if to_create { File::create(&file_path) } else { File::open(&file_path) } {
+					Ok(file) => file,
+					Err(err) => {
+						trace!(target: "privatetx", "Cannot open logs file: {}", err);
+						bail!("Cannot open logs file: {:?}", err);
+					}
+				}
+			}
+			None => {
+				warn!(target: "privatetx", "Logs path is not defined");
+				return Ok(None);
+			}
+		};
+		Ok(Some(log_file))
+	}
+}
+
+impl LogsSerializer for FileLogsSerializer {
+	fn read_logs(&self) -> Result<Vec<TransactionLog>, Error> {
+		let log_file = self.open_file(false)?;
+		match log_file {
+			Some(log_file) => {
+				let transaction_logs: Vec<TransactionLog> = match serde_json::from_reader(log_file) {
+					Ok(logs) => logs,
+					Err(err) => {
+						error!(target: "privatetx", "Cannot deserialize logs from file: {}", err);
+						bail!("Cannot deserialize logs from file: {:?}", err);
+					}
+				};
+				Ok(transaction_logs)
+			}
+			None => {
+				Ok(Vec::new())
+			}
+		}
+	}
+
+	fn flush_logs(&self, logs: &HashMap<H256, TransactionLog>) -> Result<(), Error> {
+		if logs.is_empty() {
+			// Do not create empty file
+			return Ok(());
+		}
+		let log_file = self.open_file(true)?;
+		if let Some(log_file) = log_file {
+			let mut json = serde_json::Serializer::new(log_file);
+			let mut json_array = json.serialize_seq(Some(logs.len()))?;
+			for v in logs.values() {
+				json_array.serialize_element(v)?;
+			}
+			json_array.end()?;
+		}
+		Ok(())
+	}
+}
+
+/// Timestamp source for logs
+pub trait TimestampSource: Send + Sync + 'static {
+	/// Returns current timestamp in seconds
+	fn current_timestamp(&self) -> u64;
+}
+
+/// Timesource on the base of system time
+pub struct SystemTimestamp {}
+
+impl TimestampSource for SystemTimestamp {
```

Comment:
Why not implement this directly on `SystemTime`?

I also think the trait name is not ideal something like `CurrentTime` is better because the actual implementer provides the `source`

Reply:
This wrapper was done in order to use in test. I have a test for serialization, where I need to have an exact timestamp value, that cannot be guaranteed with system timestamp

Reply:
hmm, I don't understand your answer, basically I think the trait is fine which is used to create a test helper for to serialization AFAIU.

What I meant was that you can remove the `struct SystemTimeStamp` and do this instead:

```rust
impl TimestampSource for SystemTime {
    fn timestamp(&self) -> u64 {
        self.duration_since(UNIX_EPOCH).unwrap_or_default().as_secs()
    }
}
```

But it is fine either way!

Reply:
Got it! Reworked.